lib/util/collectionEmulator.js

```diff
@@ -0,0 +1,130 @@
+/*
+ * Kuzzle, a backend software, self-hostable and ready to use
+ * to power modern apps
+ *
+ * Copyright 2015-2018 Kuzzle
+ * mailto: support AT kuzzle.io
+ * website: http://kuzzle.io
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+'use strict';
+
+const
+  INTERNAL_PREFIX = '%',
+  USER_PREFIX = '&',
+  NAME_SEPARATOR = '.';
+
+/**
+ * Get the name of an esIndex for an index + collection
+ * By default, return the name of an 'user' index.
+ * The 'internal' option can be set to return the name for an internal index
+ *
+ * @param {string} index
+ * @param {string} collection
+ * @param {object} options - internal (false)
+ *
+ * @returns {string} esIndex name (eg: '&nepali#liia')
+ */
+function getESIndex (index, collection, { internal=false } = {}) {
+  const prefix = internal
+    ? INTERNAL_PREFIX
+    : USER_PREFIX;
+
+  return `${prefix}${index}${NAME_SEPARATOR}${collection}`;
+}
+
+/**
+ * Extract the index name from esIndex name
+ *
+ * @param {string} esIndex
+ *
+ * @returns {string} index name
+ */
+function extractIndex (esIndex) {
+  return esIndex.slice(1).split(NAME_SEPARATOR)[0];
```

Comment:
This creates many needless strings. Usually I wouldn't be picky about it, but I suspect that methods in this class will be invoked many, many times, and can/will be considered as critical.

Consider extracting the index this way:
```suggestion
  return esIndex.substr(1, full.indexOf(NAME_SEPARATOR) - 1);
```

Performances comparison:

```
slice + split x 4,296,059 ops/sec ±3.51% (84 runs sampled)
indexOf + substr x 686,262,008 ops/sec ±1.36% (82 runs sampled)
```

Reply:
Also, shouldn't we check that this is a well-formed kuzzle index name, and throw otherwise?

Reply:
Yes we need something like that but I'm not quite sure what to do for now since I'm still developing elasticsearch.js adaptation.
On some cases, ES return the complete list of his indexes and some of them could have been created outside Kuzzle so we just have to skip them.
I applied you review on the performance, but since these methods are subject to change in subsequent PR, I think it's better to review them later